Import all dependencies

In [39]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings 
warnings.filterwarnings('ignore')

Reading the Data

In [40]:
taxi_df=pd.read_csv('TaxiFare.csv',header=0,index_col=0)
taxi_df.head()

amount      date_time_of_pickup  longitude_of_pickup  \
unique_id                                                         
26:21.0       4.5  2009-06-15 17:26:21 UTC           -73.844311   
52:16.0      16.9  2010-01-05 16:52:16 UTC           -74.016048   
35:00.0       5.7  2011-08-18 00:35:00 UTC           -73.982738   
30:42.0       7.7  2012-04-21 04:30:42 UTC           -73.987130   
51:00.0       5.3  2010-03-09 07:51:00 UTC           -73.968095   

           latitude_of_pickup  longitude_of_dropoff  latitude_of_dropoff  \
unique_id                                                                  
26:21.0             40.721319            -73.841610            40.712278   
52:16.0             40.711303            -73.979268            40.782004   
35:00.0             40.761270            -73.991242            40.750562   
30:42.0             40.733143            -73.991567            40.758092   
51:00.0             40.768008            -73.956655            40.783762   

           no_of_passenger  
unique_id                   
26:21.0                  1  
52:16.0                  1  
35:00.0                  2  
30:42.0                  1  
51:00.0                  1

In [41]:
taxi_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 26:21.0 to 13:14.0
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   amount                50000 non-null  float64
 1   date_time_of_pickup   50000 non-null  object 
 2   longitude_of_pickup   50000 non-null  float64
 3   latitude_of_pickup    50000 non-null  float64
 4   longitude_of_dropoff  50000 non-null  float64
 5   latitude_of_dropoff   50000 non-null  float64
 6   no_of_passenger       50000 non-null  int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 3.1+ MB


convert date_time_of _pickup into date time variable( since it is a string variable) #typecasting

In [42]:
taxi_df.date_time_of_pickup=pd.to_datetime(taxi_df.date_time_of_pickup)  # dot operator for accessing data

In [43]:
taxi_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 26:21.0 to 13:14.0
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype              
---  ------                --------------  -----              
 0   amount                50000 non-null  float64            
 1   date_time_of_pickup   50000 non-null  datetime64[ns, UTC]
 2   longitude_of_pickup   50000 non-null  float64            
 3   latitude_of_pickup    50000 non-null  float64            
 4   longitude_of_dropoff  50000 non-null  float64            
 5   latitude_of_dropoff   50000 non-null  float64            
 6   no_of_passenger       50000 non-null  int64              
dtypes: datetime64[ns, UTC](1), float64(5), int64(1)
memory usage: 3.1+ MB


Descriptive Statistics Summary

In [44]:
taxi_df.describe()

amount  longitude_of_pickup  latitude_of_pickup  \
count  50000.000000         50000.000000        50000.000000   
mean      11.364171           -72.509756           39.933759   
std        9.685557            10.393860            6.224857   
min       -5.000000           -75.423848          -74.006893   
25%        6.000000           -73.992062           40.734880   
50%        8.500000           -73.981840           40.752678   
75%       12.500000           -73.967148           40.767360   
max      200.000000            40.783472          401.083332   

       longitude_of_dropoff  latitude_of_dropoff  no_of_passenger  
count          50000.000000         50000.000000     50000.000000  
mean             -72.504616            39.926251         1.667840  
std               10.407570             6.014737         1.289195  
min              -84.654241           -74.006377         0.000000  
25%              -73.991152            40.734372         1.000000  
50%              -73.980082            40.753372         1.000000  
75%              -73.963584            40.768167         2.000000  
max               40.851027            43.415190         6.000000

negative values in amount variable, latitudes-range- -90 to 90 ,longitude ideal range is -180 to 180,
0 degree latitude and longitute values- island (we will drop this records since model may learn upon it)

In [45]:
taxi_df=taxi_df.reset_index()

In [46]:
taxi_df.head()

unique_id  amount       date_time_of_pickup  longitude_of_pickup  \
0   26:21.0     4.5 2009-06-15 17:26:21+00:00           -73.844311   
1   52:16.0    16.9 2010-01-05 16:52:16+00:00           -74.016048   
2   35:00.0     5.7 2011-08-18 00:35:00+00:00           -73.982738   
3   30:42.0     7.7 2012-04-21 04:30:42+00:00           -73.987130   
4   51:00.0     5.3 2010-03-09 07:51:00+00:00           -73.968095   

   latitude_of_pickup  longitude_of_dropoff  latitude_of_dropoff  \
0           40.721319            -73.841610            40.712278   
1           40.711303            -73.979268            40.782004   
2           40.761270            -73.991242            40.750562   
3           40.733143            -73.991567            40.758092   
4           40.768008            -73.956655            40.783762   

   no_of_passenger  
0                1  
1                1  
2                2  
3                1  
4                1

we will drop unique id variable since there are many duplicate values (it may be a user_id for a same passenger)

In [47]:
taxi_df.duplicated(subset='unique_id').sum()

46403

In [48]:
taxi_df.drop('unique_id',axis=1,inplace=True) # we reseted index since unique id var was index column we cannot drop var(at index position)
taxi_df.head()

amount       date_time_of_pickup  longitude_of_pickup  latitude_of_pickup  \
0     4.5 2009-06-15 17:26:21+00:00           -73.844311           40.721319   
1    16.9 2010-01-05 16:52:16+00:00           -74.016048           40.711303   
2     5.7 2011-08-18 00:35:00+00:00           -73.982738           40.761270   
3     7.7 2012-04-21 04:30:42+00:00           -73.987130           40.733143   
4     5.3 2010-03-09 07:51:00+00:00           -73.968095           40.768008   

   longitude_of_dropoff  latitude_of_dropoff  no_of_passenger  
0            -73.841610            40.712278                1  
1            -73.979268            40.782004                1  
2            -73.991242            40.750562                2  
3            -73.991567            40.758092                1  
4            -73.956655            40.783762                1

In [49]:
taxi_df[taxi_df.amount<=0].index

Index([2039, 2486, 10002, 13032, 27891, 28839, 36722, 42337, 47302], dtype='int64')

In [50]:
# deleting the rows where amount is negative

taxi_df.drop(taxi_df[taxi_df.amount<=0].index,axis=0,inplace=True)

In [51]:
taxi_df.shape

(49991, 7)

assign function allows creating multiple column at once(first convert into date time form) 

In [ ]:
#feature engineering on date variable

In [52]:
# extract various date time components as seperate variables
taxi_df = taxi_df.assign(hour = taxi_df.date_time_of_pickup.dt.hour, 
                         day = taxi_df.date_time_of_pickup.dt.day,
                        month = taxi_df.date_time_of_pickup.dt.month, 
                        year = taxi_df.date_time_of_pickup.dt.year, 
                        dayofweek = taxi_df.date_time_of_pickup.dt.dayofweek)

In [53]:
taxi_df.head()

amount       date_time_of_pickup  longitude_of_pickup  latitude_of_pickup  \
0     4.5 2009-06-15 17:26:21+00:00           -73.844311           40.721319   
1    16.9 2010-01-05 16:52:16+00:00           -74.016048           40.711303   
2     5.7 2011-08-18 00:35:00+00:00           -73.982738           40.761270   
3     7.7 2012-04-21 04:30:42+00:00           -73.987130           40.733143   
4     5.3 2010-03-09 07:51:00+00:00           -73.968095           40.768008   

   longitude_of_dropoff  latitude_of_dropoff  no_of_passenger  hour  day  \
0            -73.841610            40.712278                1    17   15   
1            -73.979268            40.782004                1    16    5   
2            -73.991242            40.750562                2     0   18   
3            -73.991567            40.758092                1     4   21   
4            -73.956655            40.783762                1     7    9   

   month  year  dayofweek  
0      6  2009          0  
1      1  2010          1  
2      8  2011          3  
3      4  2012          5  
4      3  2010          1

In [54]:
taxi_df.drop('date_time_of_pickup',axis=1,inplace=True)
taxi_df.head()

amount  longitude_of_pickup  latitude_of_pickup  longitude_of_dropoff  \
0     4.5           -73.844311           40.721319            -73.841610   
1    16.9           -74.016048           40.711303            -73.979268   
2     5.7           -73.982738           40.761270            -73.991242   
3     7.7           -73.987130           40.733143            -73.991567   
4     5.3           -73.968095           40.768008            -73.956655   

   latitude_of_dropoff  no_of_passenger  hour  day  month  year  dayofweek  
0            40.712278                1    17   15      6  2009          0  
1            40.782004                1    16    5      1  2010          1  
2            40.750562                2     0   18      8  2011          3  
3            40.758092                1     4   21      4  2012          5  
4            40.783762                1     7    9      3  2010          1

In [ ]:
#trying to see 0 value in lat and longitude variable


In [ ]:
# drop those vars where a observation has 0 value as a record

In [55]:
taxi_df.latitude_of_pickup.value_counts()

latitude_of_pickup
0.000000     955
41.366138     18
40.756007     10
40.763975      8
40.757500      7
            ... 
40.725295      1
40.768578      1
40.740649      1
40.772588      1
40.768434      1
Name: count, Length: 36587, dtype: int64

In [58]:
taxi_df.drop(taxi_df[(taxi_df.latitude_of_pickup==0) | (taxi_df.longitude_of_pickup==0) |(taxi_df.latitude_of_dropoff==0)|(taxi_df.longitude_of_dropoff==0)].index,inplace=True)

In [59]:
taxi_df.shape

(48992, 11)

In [60]:
# data entry mistake (401 is max value in latitude pickup and beyound 90 for longitude)
# replace that max value 401

taxi_df.latitude_of_pickup.sort_values(ascending=True)

8443     -74.006893
10091    -74.006210
31847    -74.004027
41872    -73.995000
9548     -73.992947
            ...    
42869     41.523217
43305     41.650000
35211     42.160275
28274     43.098708
5686     401.083332
Name: latitude_of_pickup, Length: 48992, dtype: float64

In [61]:
taxi_df.latitude_of_pickup.replace( 401.083332,40.1083332,inplace=True)

In [62]:
taxi_df.latitude_of_pickup.value_counts().index.sort_values(ascending=True)

Index([-74.006893,  -74.00621, -74.004027,    -73.995, -73.992947,  -73.99184,
       -73.988467, -73.987585, -73.987307, -73.986968,
       ...
          41.0091,   41.03249,  41.035688,  41.150487,  41.366138,  41.391042,
        41.523217,      41.65,  42.160275,  43.098708],
      dtype='float64', name='latitude_of_pickup', length=36565)

In [63]:
taxi_df.describe()

amount  longitude_of_pickup  latitude_of_pickup  \
count  48992.000000         48992.000000        48992.000000   
mean      11.362482           -73.927624           40.703940   
std        9.655941             2.308123            2.281165   
min        0.010000           -75.423848          -74.006893   
25%        6.000000           -73.992297           40.736466   
50%        8.500000           -73.982115           40.753412   
75%       12.500000           -73.968355           40.767755   
max      200.000000            40.783472           43.098708   

       longitude_of_dropoff  latitude_of_dropoff  no_of_passenger  \
count          48992.000000         48992.000000     48992.000000   
mean             -73.920886            40.703647         1.669354   
std                2.396053             2.289503         1.290292   
min              -84.654241           -74.006377         0.000000   
25%              -73.991327            40.735862         1.000000   
50%              -73.980466            40.754085         1.000000   
75%              -73.965265            40.768479         2.000000   
max               40.851027            43.415190         6.000000   

               hour           day         month          year     dayofweek  
count  48992.000000  48992.000000  48992.000000  48992.000000  48992.000000  
mean      13.483283     15.666272      6.273147   2011.739141      3.031107  
std        6.503869      8.659639      3.461245      1.864828      1.956408  
min        0.000000      1.000000      1.000000   2009.000000      0.000000  
25%        9.000000      8.000000      3.000000   2010.000000      1.000000  
50%       14.000000     16.000000      6.000000   2012.000000      3.000000  
75%       19.000000     23.000000      9.000000   2013.000000      5.000000  
max       23.000000     31.000000     12.000000   2015.000000      6.000000

In [ ]:
#haversine function for finding actual distance

In [64]:
#! pip install haversine

  Obtaining dependency information for haversine from https://files.pythonhosted.org/packages/5b/f1/b7274966f0b5b665d9114e86d09c6bc87d241781d63d8817323dcfa940c6/haversine-2.8.1-py2.py3-none-any.whl.metadata


In [65]:
from haversine import haversine, Unit
#df.apply(lambda x: func(x['col1'],x['col2']),axis=1)
#haversine((lat1,long1),(lat2,long2))
taxi_df['travel_dist_km']=taxi_df.apply(lambda row: haversine((row["latitude_of_pickup"],row["longitude_of_pickup"]),
                                                              (row["latitude_of_dropoff"],row["longitude_of_dropoff"]),
                                                              unit=Unit.KILOMETERS),axis=1)
taxi_df['travel_dist_km']=taxi_df['travel_dist_km'].round(2)

In [66]:
taxi_df.head()

amount  longitude_of_pickup  latitude_of_pickup  longitude_of_dropoff  \
0     4.5           -73.844311           40.721319            -73.841610   
1    16.9           -74.016048           40.711303            -73.979268   
2     5.7           -73.982738           40.761270            -73.991242   
3     7.7           -73.987130           40.733143            -73.991567   
4     5.3           -73.968095           40.768008            -73.956655   

   latitude_of_dropoff  no_of_passenger  hour  day  month  year  dayofweek  \
0            40.712278                1    17   15      6  2009          0   
1            40.782004                1    16    5      1  2010          1   
2            40.750562                2     0   18      8  2011          3   
3            40.758092                1     4   21      4  2012          5   
4            40.783762                1     7    9      3  2010          1   

   travel_dist_km  
0            1.03  
1            8.45  
2            1.39  
3            2.80  
4            2.00

we dont need that four columns now

In [67]:
taxi_df.columns

Index(['amount', 'longitude_of_pickup', 'latitude_of_pickup',
       'longitude_of_dropoff', 'latitude_of_dropoff', 'no_of_passenger',
       'hour', 'day', 'month', 'year', 'dayofweek', 'travel_dist_km'],
      dtype='object')

In [68]:
taxi_df.drop(['longitude_of_pickup', 'latitude_of_pickup',
       'longitude_of_dropoff', 'latitude_of_dropoff'],axis=1,inplace=True)
taxi_df.head()

amount  no_of_passenger  hour  day  month  year  dayofweek  travel_dist_km
0     4.5                1    17   15      6  2009          0            1.03
1    16.9                1    16    5      1  2010          1            8.45
2     5.7                2     0   18      8  2011          3            1.39
3     7.7                1     4   21      4  2012          5            2.80
4     5.3                1     7    9      3  2010          1            2.00

now basic preprocessing part is done

Creating X and Y

In [69]:
X=taxi_df.values[:,1:]
Y=taxi_df.values[:,0]

Scaling the data

In [70]:
from sklearn.preprocessing import StandardScaler
# create an object
scaler=StandardScaler()

scaler.fit(X)
X=scaler.transform(X)
#X=scalar.fit_transform(X)

Splitting the data using train_test_split

In [71]:
from sklearn.model_selection import train_test_split
 
#Split the data into test and train
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, 
                                                    random_state=10)  

In [72]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(34294, 7)
(14698, 7)
(34294,)
(14698,)


## Decision Tree Regressor

In [74]:
from sklearn.tree import DecisionTreeRegressor

# create model object 
model_DecisionTree=DecisionTreeRegressor(random_state=10,criterion='squared_error') #Squared error id SSE (lower is the SSE higher is significance of variable)

# fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)

Y_pred=model_DecisionTree.predict(X_test)


print(Y_pred)


[36.5  5.   8.9 ... 11.5 15.5 11. ]


Accuracy on Training data using score function

In [75]:
model_DecisionTree.score(X_train,Y_train)

0.9999926735062961

Evaluating the model

In [77]:

from sklearn.metrics import r2_score,mean_squared_error
import numpy as np
 
r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)
 
rmse=np.sqrt(mean_squared_error(Y_test,Y_pred)) 
print("RMSE:",rmse)
 
adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)  # formula of adj R2 shape(shape-no of ind vars)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.41934728538014276
RMSE: 7.293701662626026
Adj R-square: 0.4192643078976518


it is overfitting case (training part more accuracy ,testing part(R2) low accuracy(0.9999926735062961 actual,0.41934728538014276 test)

this tree is overfitted
we will prune the tree

In [ ]:
#we can use gridsearchcv also for tuning


In [78]:
from sklearn.tree import DecisionTreeRegressor

# create model object 
model_DecisionTree=DecisionTreeRegressor(random_state=10,max_depth=5,min_samples_leaf=5) #Squared error id SSE (lower is the SSE higher is significance of variable)

# fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)

Y_pred=model_DecisionTree.predict(X_test)


print(Y_pred)


[26.31727147  5.50436867  9.68327851 ... 10.02815744 18.94743557
 10.02815744]


In [79]:
model_DecisionTree.score(X_train,Y_train)

0.7461592894163125

In [80]:
from sklearn.metrics import r2_score,mean_squared_error
import numpy as np
 
r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)
 
rmse=np.sqrt(mean_squared_error(Y_test,Y_pred)) 
print("RMSE:",rmse)
 
adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)  # formula of adj R2 shape(shape-no of ind vars)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.7861636482405271
RMSE: 4.426194895667306
Adj R-square: 0.7861330902121441


since score function and R squared value (for testing ) are close to each other implies that is a good model
R2 and adj R2 close to each other good
RMSE is lower than previous model and after prunning we are getting a better model

In [83]:
print((list(zip(taxi_df.columns[1:],
               model_DecisionTree.feature_importances_))))

[('no_of_passenger', 0.0), ('hour', 0.0016419068255963175), ('day', 0.0), ('month', 0.0), ('year', 0.02160810007674775), ('dayofweek', 0.0010219225126868697), ('travel_dist_km', 0.9757280705849691)]


travel_dist_km is most significant varible (it is used many times to build a tree)

In [84]:
from sklearn import tree
with open(r"model_DecisionTree.txt", "w") as f:
    
    f = tree.export_graphviz(model_DecisionTree, feature_names=taxi_df.columns[1:],
                             out_file=f)
    
#generate the file and upload the code in webgraphviz.com to plot the decision tree

## Random Forest

In [85]:

from sklearn.ensemble import RandomForestRegressor
 
model_DecisionTree=RandomForestRegressor(n_estimators=100,# how many trees to grow
                                          random_state=10, bootstrap=True,#with replacements
                                         )
 
#fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)
 
Y_pred=model_DecisionTree.predict(X_test)
print(Y_pred)

[28.6451  5.61    8.26   ... 10.465  19.145   8.615 ]


In [86]:
model_DecisionTree.score(X_train,Y_train)

0.9607304210830027

Evaluation

In [87]:
from sklearn.metrics import r2_score,mean_squared_error
import numpy as np
 
r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)
 
rmse=np.sqrt(mean_squared_error(Y_test,Y_pred)) 
print("RMSE:",rmse)
 
adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)  # formula of adj R2 shape(shape-no of ind vars)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.768891122940369
RMSE: 4.601485719098559
Adj R-square: 0.768858096602393


there is difference between score function and R2 (hence overfitting 20 % difference)

In [89]:

from sklearn.ensemble import RandomForestRegressor
 
model_DecisionTree=RandomForestRegressor(n_estimators=100,# how many trees to grow
                                          random_state=10,max_depth=5,min_samples_leaf=5, bootstrap=True,#with replacements
                                         )
 
#fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)
 
Y_pred=model_DecisionTree.predict(X_test)
print(Y_pred)

[26.58723949  5.48362388  9.45751295 ... 10.79499642 19.35004592
  9.80453887]


In [90]:
model_DecisionTree.score(X_train,Y_train)

0.7523209862686894

In [91]:
from sklearn.metrics import r2_score,mean_squared_error
import numpy as np
 
r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)
 
rmse=np.sqrt(mean_squared_error(Y_test,Y_pred)) 
print("RMSE:",rmse)
 
adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)  # formula of adj R2 shape(shape-no of ind vars)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.7924955610078093
RMSE: 4.36017038391129
Adj R-square: 0.7924659078338556


pruned RF is better than default RF

In [92]:
print((list(zip(taxi_df.columns[1:],
               model_DecisionTree.feature_importances_))))

[('no_of_passenger', 0.0002065736225883579), ('hour', 0.0034770085513773407), ('day', 0.0003102995182481823), ('month', 0.0008841426530119284), ('year', 0.021811050686318393), ('dayofweek', 0.0005923403515313504), ('travel_dist_km', 0.9727185846169245)]


In [95]:
#we will implement pruned extra tree directly

#predicting using the Random_Forest_Classifier
from sklearn.ensemble import ExtraTreesRegressor
 
model_DecisionTree=ExtraTreesRegressor(n_estimators=100,# how many trees to grow
                                          random_state=10,max_depth=5,min_samples_leaf=5, bootstrap=True,#with replacements
                                         )
 
#fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)
 
Y_pred=model_DecisionTree.predict(X_test)
print(Y_pred)

[11.25343918 10.2137769   9.22684473 ... 11.57628794 16.57729021
 10.89544267]


In [96]:
model_DecisionTree.score(X_train,Y_train)

0.39853627129784774

In [97]:
from sklearn.metrics import r2_score,mean_squared_error
import numpy as np
 
r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)
 
rmse=np.sqrt(mean_squared_error(Y_test,Y_pred)) 
print("RMSE:",rmse)
 
adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)  # formula of adj R2 shape(shape-no of ind vars)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.4291842698232019
RMSE: 7.231655533375233
Adj R-square: 0.42910269808322077


pruned Random Forest is giving more accuracy out of all the models
any future prediction we do we will do that on Random Forest model

KNN does not work in training part but works only on test data hence it is called lazy model